### ISSUE TITLE
Smoother disposing

### ISSUE THREAD

**Issue by mfagerlund** [dotnet/TorchSharp#450]
Hi, I've created a fork where I demonstrate a simpler form of disposing in TorchSharp. 

The idea is to wrap code in usings of DisposeScope , any  tensor created on a device that has a DisposeScope will be added  to a list of tensors  that are later disposed.

It can be used  thusly;

        [Test]
        public void MemoryManagementExample()
        {
            torch.Tensor sum;
            _testOutputHelper.WriteLine($"Total tensors: {torch.Tensor.TotalCount}"); // Writes 0
            using (var d = torch.CPU.NewDisposeScope()) {
                var predictedLabels = torch.rand(5, 5);
                var labels = torch.tensor(new int[] { 1, 0, 1, 0, 1 });
                var am = predictedLabels.argmax(1);
                var eq = am == labels;
                _testOutputHelper.WriteLine($"Total tensors temporary dispose: {torch.Tensor.TotalCount}"); // Writes 4
                // If the process is long, you can dispose mid stream;
                d.DisposeEverythingBut(eq);
                // This last value is excluded from the DisposeScope, so it can be used outside.
                sum = d.Exclude(eq.sum());
                _testOutputHelper.WriteLine($"Total tensors before dispose: {torch.Tensor.TotalCount}"); // Writes 2
            }

            _testOutputHelper.WriteLine($"Total tensors after dispose: {torch.Tensor.TotalCount}"); // Writes 1
            _testOutputHelper.WriteLine($"Tensors disposed: {DisposeScopeManager.DisposedTensorCount}"); // Writes 4
            _testOutputHelper.WriteLine($"sum:{sum.ToSingle()}");
        }

You can have a look see here: https://github.com/mfagerlund/TorchSharp , if this could be useful  to others, I'd be happy to create a pull request. Any thoughts on the naming and such is more than welcome,

**Comment by mfagerlund** [comment]
Here's another example, where the test TestTraining.TestTraining1() leaves 184 dangling tensors, that are all caught when  running inside a DisposeScope

        [Fact]
        public void DisposeScopeWorksForTestTraining1()
        {
            var count = torch.Tensor.TotalCount;
            // No dispose scope
            var testTraining = new TestTraining();
            testTraining.TestTraining1();
            _testOutputHelper.WriteLine($"Undisposed Tensors without DisposeScope: {torch.Tensor.TotalCount - count}");

            count = torch.Tensor.TotalCount;
            using (torch.CPU.NewDisposeScope()) {
                testTraining.TestTraining1();
            }

            _testOutputHelper.WriteLine($"Undisposed Tensors with DisposeScope: {torch.Tensor.TotalCount - count}");
            Assert.Equal(torch.Tensor.TotalCount, count);
        }

**Comment by mfagerlund** [comment]
Should I have posted this in discussions instead?

**Comment by lostmsu** [comment]
This sounds like a very good idea to me. Until C# supports some form of RAII, we might want to emulate it. Currently, the requirement to manually dispose tensors is very painful.

Another option I was thinking about was to modify .NET itself to call finalizers for temp variables that do not leave some scope.

**Comment by NiklasGustafsson** [comment]
@mfagerlund -- anything we can do to simplify disposal is going to be highly valuable. I've spent a ridiculous amount of energy on pulling out temporaries and adding 'usings' and the only result is hard-to-read code. 

That said, I think it's worth having more of a discussion before generating a PR.

For example, one of the questions I have is about nesting of these scopes -- I may suppress disposal in one scope, then return the instance to a calling function where I don't want it suppressed, or I may. Does suppression (DisposeEverythingBut) transfer ownership to the parent scope?

The scope notion is thread-local, right?

**Comment by NiklasGustafsson** [comment]
@mfagerlund -- why 'torch.CPU' as the containing namespace? I'd think this either belongs in 'torch' or under "torch.tensor.' To me, "CPU" suggests that it's not applicable to CUDA-hosted tensors, which I presume is not what you are suggesting.

**Comment by mfagerlund** [comment]
@NiklasGustafsson torch.CPU - Where is that the containing namespace? If it is, that's a mistake. But it looks to me like DisposeScopeManager is in  the TorchSharp namespace? 

  ```
namespace TorchSharp
  {
      /// <summary>
      /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
      /// DisposeManager isn't "thread safe", since all CPU Devices are the same, the Dispose manager
      /// would pick up tensors from other threads. Therefore, each thread gets it's own DisposeScopeManager.
      /// </summary>
      public class DisposeScopeManager
```

I'd gladly move it to the torch namespace, thought that's a class and not a namespace?

```
    public static partial class torch
    {
```

"Does suppression (DisposeEverythingBut) transfer ownership to the parent scope?"

It doesn't suppress it at all, only temporarily if your scope is long lived. Exclude suppresses from the current scope and moves it to the parent. ExcludeGlobally detaches the tensor from the process altogether.

All these are super relevant questions and the naming should match what users of TorchSharp would typically expect!

"The scope notion is thread-local, right?"

Yes, I couldn't figure out any other way of dealing with it. Especially the CPU device, there's only one and stealing tensors from other thread seems like a bad idea...

"I would also like some sense of the runtime cost."

I would love to use a HashSet instead of a list, but it turns out that sometimes the tensors are disposed even though I don't do it manually. Once a tensor is disposed, it's handle becomes invalid and any reference to Equals (which both List.Remove and HashSet does) fails;

```
  public bool Equals(Tensor target)
            {
                var res = THSTensor_equal(Handle, target.Handle);
                torch.CheckForErrors();
                return res;
            }
```

Possibly, if  the handle has been disposed, a different equals method could be used? Then we could use a HashSet instead.

**Comment by mfagerlund** [comment]
@NiklasGustafsson I think the best low cost version of this would be if TorchSharp would implement a static hook in the tensor constructor;

```
            public Action<Tensor> ConstructorAction { get; set; }
            internal Tensor(IntPtr handle)
            {
                this.handle = handle;
                System.Threading.Interlocked.Increment(ref _totalCount);
                _peakCount = Math.Max(_totalCount, _peakCount);
                ConstructorAction?.Invoke(this);
            }
```

That way, the rest of this stuff could be injected, until it's stable enough to be accepted into the master.

The two methods I introduced, NewDisposeScope for Tensor and Device, could be implemented as extension methods in an external package.

/m

**Comment by mfagerlund** [comment]
> This sounds like a very good idea to me. 

Check it out, suggest any modifications you'd like to see!

**Comment by NiklasGustafsson** [comment]
>> @NiklasGustafsson torch.CPU - Where is that the containing namespace?

From the example you lead with:

```C#
using (var d = torch.CPU.NewDisposeScope()) {

```

**Comment by mfagerlund** [comment]
Yep, exactly, it's the device. You can either start them on a tensor or a device - but in the end they both fall back to the device_type and device_index.

**Comment by NiklasGustafsson** [comment]
I would advice against associating it with the device -- the orchestration logic is always running on the CPU, anyway, and the dynamic lifetimes are not really dependent on the device that the tensor lives on.

I suggest treating it like 'no_grad()':

```C#
using (torch.no_grad()) { ... }
```

**Comment by mfagerlund** [comment]
Hm.... yes. that would make sense. Ignore who holds the tensor data, if it's out of scope, it's gotta go. I like it.

How about the naming convention, when do you follow pytorch style and when do you follow c#?

`using(torch.dispose_scope())`

vs

`using(torch.CreateDisposeScope())`

?

**Comment by NiklasGustafsson** [comment]
Assembling a few things in one place:

>> I'd gladly move it to the torch namespace, thought that's a class and not a namespace?

Yes, that's a workaround the fact that you can't put methods directly in namespaces. We're essentially treating static classes as namespaces. Ugly? For sure. Pragmatic? Yes.

>> I think the best low cost version of this would be if TorchSharp would implement a static hook in the tensor constructor;

It's not obvious to me why that is lost-cost per se, but it's a good extension mechanism, regardless. For composition purposes, we should probably have something more like a registry of hooks, rather than a single (static) property with a public setter.

>> Possibly, if the handle has been disposed, a different equals method could be used?

Presumably, if you use a list, the only scenario where this can come up is if you are trying to exclude a tensor that has already been disposed. That seems like an error to me. For HashSet, I understand that it would be a problem also for insert. I don't know if list insert is constant-time, but it would upset me to learn that it isn't.

>> How about the naming convention, when do you follow pytorch style and when do you follow c#?

The basic principle is to follow PyTorch style when it's related to an API that we copied from Python, .NET style otherwise. You can argue about the wisdom of that, and I could be persuaded that we should use PyTorch style for everything, but that would still leave things like ToString() and Dispose(), which you can't get rid of.

**Comment by NiklasGustafsson** [comment]
Let me just say that I really like this idea. In fact, it's the fact that it's so very useful that it's important to nail down the design details and the runtime costs, because using this is something I can see us recommend as a best practice.

**Comment by mfagerlund** [comment]
> > > I'd gladly move it to the torch namespace, thought that's a class and not a namespace?
> 
> Yes, that's a workaround the fact that you can't put methods directly in namespaces. We're essentially treating static classes as namespaces. Ugly? For sure. Pragmatic? Yes.

I agree with all your design choices, for the reasons you've given. This way, it's easy to convert pytorch examples.
 
> > > I think the best low cost version of this would be if TorchSharp would implement a static hook in the tensor constructor;
> 
> It's not obvious to me why that is lost-cost per se, but it's a good extension mechanism, regardless. For composition purposes, we should probably have something more like a registry of hooks, rather than a single (static) property with a public setter.

"Low cost" out of a "changes to TorchSharp" perspective, give me the hook and I can build it in an outside package. But a more sophisticated callback system would be nice.

One way of looking at it is that TorchSharp is huge and unwieldly to deploy very often. If there was an external package, I call 
 mine torchu, that could iterate faster, then that could be of benefit to the community. Full of helper stuff that doesn't fit neatly into the core package, but none the less useful.

Having said that, there are a couple of changes beyond the hook that I'd suggest. Specifically protecting against double dispose and making sure that the Equals method doesn't throw exceptions for disposed tensors.

> > > Possibly, if the handle has been disposed, a different equals method could be used?
> 
> Presumably, if you use a list, the only scenario where this can come up is if you are trying to exclude a tensor that has already been disposed. 

I'm not sure where it happens, but some of the tensors are disposed before the end of my scope. Probably not a bug, just that something on the inside is careful to not leave dangling tensors. Should be easy enough to cope with.

Tensors could call the DisposeScope in its Dispose to let the scope know that it's finished.

> That seems like an error to me. For HashSet, I understand that it would be a problem also for insert. I don't know if list insert is constant-time, but it would upset me to learn that it isn't.

List insert is constant time if there's space, otherwise the list doubles in size and then it's constant again. But removing stuff becomes slow in a list. I'll see if I can fix the issue.

> > > How about the naming convention, when do you follow pytorch style and when do you follow c#?
> 
> The basic principle is to follow PyTorch style when it's related to an API that we copied from Python, .NET style otherwise. You can argue about the wisdom of that, and I could be persuaded that we should use PyTorch style for everything, but that would still leave things like ToString() and Dispose(), which you can't get rid of.

Great, I'll stick to that.

**Comment by mfagerlund** [comment]
> Let me just say that I really like this idea. In fact, it's the fact that it's so very useful that it's important to nail down the design details and the runtime costs, because using this is something I can see us recommend as a best practice.

Great!

**Comment by mfagerlund** [comment]
I updated my repo so that the torch class contains the NewDisposeScope method and I created callbacks in both  the constructor and dispose of tensors, so we can keep track of them and remove them from the scope if they're disposed externally. Oh, and I added "PeakMemory" to the tensor, to keep track of memory usage.

I just some tests with MNIST, comparing DisposeScope to of explicit usings. With no usings and no DisposeScope, MNIST crashes from overuse of memory. 

**With explicit usings**: End-of-epoch memory use: 137960, Peak Memory: 1 988 647 952

**With DisposeScope**: End-of-epoch memory use: 137960, Peak Memory: 1 988 647 936

So exactly the same, and the DisposeScope version can be written as

                using (var scope = torch.NewDisposeScope()) {
                    var l1 = relu1.forward(conv1.forward(input));
                    var l2 = dropout1.forward(pool1.forward(relu2.forward( conv2.forward(l1))));
                    var x = flatten.forward(l2);

                    var l33 = dropout2.forward(relu3.forward(fc1.forward(x)));
                    var l41 = fc2.forward(l33);

                    return scope.Exclude(logsm.forward(l41));
                }

Seems to work.

**Comment by NiklasGustafsson** [comment]
Okay, I'll play around with this later today. I suspect that there are going to be some things around the edges that will need tweaking. I wonder if it's possible to implement 'Exclude()' in a way that doesn't require removing it from the list, kind of a 'logical delete' to borrow a term from relational databases.

>> If there was an external package, I call mine torchu, that could iterate faster, then that could be of benefit to the community.

Sure, and I think the kinds of things you showed in the discussion belong in their own library, maybe it's something to integrate later. This one thing, though, goes to the very core of how to use TorchSharp, addressing the most irritating (in my opinion) user experience issue we're facing. Therefore, I think this belongs in TorchSharp proper.

That said, a hooking mechanism has so many possible uses, for diagnostics, debugging, etc., so it's worth adding regardless as long as it can be done without adversely affecting the runtime performance of tensor create/dispose.

>> I'm not sure where it happens, but some of the tensors are disposed before the end of my scope. Probably not a bug, just that something on the inside is careful to not leave dangling tensors. Should be easy enough to cope with.

Hmmm. That is very odd. I'll want to take a look at that. Basically, I don't see how that should ever happen, but I could be overlooking something.

The double dispose -- I want to understand more about what protection you are looking for. Right now, once the handle is Zero, dispose is a no-op.

Most important to me is to figure out is the nesting of these dynamic scopes and whether there's some odd interaction with multi-threading. I don't think this introduces anything new w.r.t. threading, but it's worth thinking about it.

**Comment by mfagerlund** [comment]
> I wonder if it's possible to implement 'Exclude()' in a way that doesn't require removing it from the list, kind of a 'logical delete' to borrow a term from relational databases.

I modified the equals method of Tensor so it's safe  to use even after the tensor has been disposed, so I can use a hashset instead of a list which is much faster. Not sure we need to worry about the deletes?

> Sure, and I think the kinds of things you showed in the discussion belong in their own library, maybe it's something to integrate later. This one thing, though, goes to the very core of how to use TorchSharp, addressing the most irritating (in my opinion) user experience issue we're facing. Therefore, I think this belongs in TorchSharp proper.

Sounds reasonable.
 
> Hmmm. That is very odd. I'll want to take a look at that. Basically, I don't see how that should ever happen, but I could be overlooking something.

I'll try to pinpoint when it happens, I may  have imagined the entire thing ;)
 
> The double dispose -- I want to understand more about what protection you are looking for. Right now, once the handle is Zero, dispose is a no-op.

Yeah, I think the issue I was having was that Equals became dangerous after a dispose - the dispose itself was always safe.

> Most important to me is to figure out is the nesting of these dynamic scopes and whether there's some odd interaction with multi-threading. I don't think this introduces anything new w.r.t. threading, but it's worth thinking about it.

Let me know what you find!

**Comment by NiklasGustafsson** [comment]
@mfagerlund -- please submit a PR, and we'll use the review process to finalize this.

**Comment by mfagerlund** [comment]
I found an issue, and it may have been the unexpected dispose I was talking about; the xUnit tests run in parallell on multiple threads.

**Comment by NiklasGustafsson** [comment]
Okay. No need to close the issue for that. Let's continue to use this space to track progress.

One initial comment on the code (I'm looking at your DisposeScope branch) -- it still seems to be associating scopes with devices, which I don't think it should. It's really just a stack of scopes for each thread.

Also, I don't see your 'PeakMemory' property in the code. I had that in at one point, counting the size of tensors, but it's not reliable, since there are so many situations where .NET Tensor instances refer to the same underlying storage, and I got a lot of double-counting. In particular, it will happen with any API that creates a tensor view, Tensor.alias(), and all the in-place operators. If that's how you have it implemented, I think it will overestimate the amount of memory in use.

**Comment by mfagerlund** [comment]
I'm having issues with GitHub, I have a lot of changes don't seem to be able to push. Will  get back once it's working. I hear you about the threads, I'll have a look at find a better way of handling it.

**Comment by NiklasGustafsson** [comment]
Okay, no pressure. :-)

**Comment by mfagerlund** [comment]
There's some new fresh code there now, no reference to Device anywhere! And I didn't mean to close  the issue, I didn't even realize that I had done that ;)
